File: jvm/tests/environmental/bme680/Bme680Test.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 22+
//JAVA_OPTIONS --enable-native-access=ALL-UNNAMED
//DEPS it.uhde:periph-transport:1.0-SNAPSHOT
//DEPS it.uhde:periph-java:1.0-SNAPSHOT

import it.uhde.periph.transport.I2CTransport;
import it.uhde.periph.chips.environmental.Bme680Full;

public class Bme680Test {

    static int passed = 0;
    static int failed = 0;

    static void checkTrue(String label, boolean condition) {
        if (condition) { System.out.println("PASS " + label); passed++; }
        else           { System.out.println("FAIL " + label); failed++; }
    }

    public static void main(String[] args) throws Exception {
        int bus  = Integer.parseInt(System.getenv().getOrDefault("I2C_BUS",  "1"));
        int addr = Integer.decode(System.getenv().getOrDefault("I2C_ADDR", "0x76"));

        try (var transport = new I2CTransport(bus, addr)) {

            var sensor = new Bme680Full(transport);

            // chip_id — 0x61 = BME680
            int id = sensor.chipId();
            checkTrue("chip_id: chipId() == 0x61", id == 0x61);

            // temperature_range — must be in sensor operating range
            double t = sensor.temperature();
            checkTrue("temperature_range: temperature() in range [-20, 85] °C",
                      t >= -20.0 && t <= 85.0);

            // pressure_range — valid range per datasheet
            double p = sensor.pressure();
            checkTrue("pressure_range: pressure() in range [300, 1100] hPa",
                      p >= 300.0 && p <= 1100.0);

            // humidity_range — valid range per datasheet
            double h = sensor.humidity();
            checkTrue("humidity_range: humidity() in range [0, 100] %RH",
                      h >= 0.0 && h <= 100.0);

            // gas_resistance — must return a positive value
            double g = sensor.gasResistance();
            checkTrue("gas_resistance: gasResistance() > 0", g > 0.0);

            // default_oversampling — configure() with defaults accepted
            sensor.configure(Bme680Full.OSRS_X2, Bme680Full.OSRS_X16,
                             Bme680Full.OSRS_X1, Bme680Full.MODE_FORCED,
                             Bme680Full.FILTER_0);
            checkTrue("default_oversampling: configure() accepted", true);

            // set_oversampling — must be accepted without exception
            sensor.setOversampling(Bme680Full.OSRS_X4, Bme680Full.OSRS_X4,
                                   Bme680Full.OSRS_X4);
            checkTrue("set_oversampling: setOversampling() accepted", true);

            // read_all — returns double[4] with valid ranges
            double[] all = sensor.readAll();
            checkTrue("read_all: array length == 4", all.length == 4);
            checkTrue("read_all: temperature in range [-20, 85] °C",
                      all[0] >= -20.0 && all[0] <= 85.0);
            checkTrue("read_all: pressure in range [300, 1100] hPa",
                      all[1] >= 300.0 && all[1] <= 1100.0);
            checkTrue("read_all: humidity in range [0, 100] %RH",
                      all[2] >= 0.0 && all[2] <= 100.0);

            // reset — must complete without exception and keep sensor functional
            sensor.reset();
            checkTrue("reset: reset() accepted", true);

            double tAfter = sensor.temperature();
            checkTrue("reset: temperature() after reset in range [-20, 85] °C",
                      tAfter >= -20.0 && tAfter <= 85.0);

        }

        System.out.printf("===DONE: %d passed, %d failed===%n", passed, failed);
        System.exit(failed == 0 ? 0 : 1);
    }
}

File: nodejs/tests/environmental/bme680_test.js
'use strict';
const { I2CTransport } = require('../../packages/periph/src/transport/i2c');
const { BME680Minimal, BME680Full } = require('../../packages/periph/src/chips/environmental/bme680');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR  || '0x76', 16);

let passed = 0, failed = 0;

const transport = new I2CTransport(I2C_BUS, I2C_ADDR);
const bme = new BME680Minimal(transport);

const t = bme.temperature();
if (t >= -40 && t <= 85) { console.log('PASS temperature_range'); passed++; }
else { console.log('FAIL temperature_range: got ' + t); failed++; }

const p = bme.pressure();
if (p >= 300 && p <= 1100) { console.log('PASS pressure_range'); passed++; }
else { console.log('FAIL pressure_range: got ' + p); failed++; }

const h = bme.humidity();
if (h >= 0 && h <= 100) { console.log('PASS humidity_range'); passed++; }
else { console.log('FAIL humidity_range: got ' + h); failed++; }

const bmeFull = new BME680Full(transport);
if (bmeFull._osrsT === 1 && bmeFull._osrsP === 1 && bmeFull._osrsH === 1) { console.log('PASS default_oversampling'); passed++; }
else { console.log('FAIL default_oversampling'); failed++; }

bmeFull.setOversampling(BME680Full.OSRS_X4, BME680Full.OSRS_X2, BME680Full.OSRS_X1);
if (bmeFull._osrsT === 3 && bmeFull._osrsP === 2 && bmeFull._osrsH === 1) { console.log('PASS set_oversampling'); passed++; }
else { console.log('FAIL set_oversampling'); failed++; }

const cid = bmeFull.chipId();
if (cid === 0x61) { console.log('PASS chip_id'); passed++; }
else { console.log('FAIL chip_id: expected 0x61, got 0x' + cid.toString(16)); failed++; }

const r = bmeFull.readAll();
if (r.temperature >= -40 && r.temperature <= 85 && r.pressure >= 300 && r.pressure <= 1100 && r.humidity >= 0 && r.humidity <= 100) { console.log('PASS read_all'); passed++; }
else { console.log('FAIL read_all: T=' + r.temperature + ' P=' + r.pressure + ' H=' + r.humidity); failed++; }

bmeFull.reset();
console.log('PASS reset'); passed++;

transport.close();
console.log('===DONE: ' + passed + ' passed, ' + failed + ' failed===');
process.exit(failed === 0 ? 0 : 1);
